source/system/structs/HashMap.ooc

```diff
@@ -95,262 +60,149 @@ acX31Hash: func <K> (key: K) -> SizeT {
-				memcpy(newEntry@ value, value, V size)
-
-				currentPointer@ next = newEntry
+	remove: func ~nulldefault (key: K) -> V {
+		this remove(key, null)
+	}
+	remove: func (key: K, defaultValue: V) -> V {
+		result: V = defaultValue
+		hashkey := This hash(key) % this capacity
+		entry := this _buckets[hashkey]
+		if (entry != null) {
+			if (this _keyEquals(key, entry _key)) {
+				next := entry _next
+				entry _next = null
+				result = entry _value
+				entry free()
+				this _buckets[hashkey] = next
+				this _count -= 1
 			} else {
-				entry key = calloc(1, K size)
-				memcpy(entry key, key, K size)
-
-				entry value = calloc(1, V size)
-				memcpy(entry value, value, V size)
-
-				entry next = null
-
-				this buckets[hash] = entry
+				prev := entry
+				entry = entry _next
+				while (entry != null && !this _keyEquals(key, entry _key)) {
+					prev = entry
+					entry = entry _next
+				}
+				if (entry != null) {
+					prev _next = entry _next
+					entry _next = null
+					result = entry _value
+					entry free()
+					this _count -= 1
+				}
 			}
-			this _count += 1
-			if ((this _count as Float / this capacity as Float) > 0.75)
-				this resize(this _count * (this _count > 50000 ? 2 : 4))
 		}
-	}
-	get: func (key: K) -> V {
-		entry: HashEntry
-		result: V = null
-		if (this getEntry(key, entry&))
-			result = entry value as V
 		result
 	}
 	contains: func (key: K) -> Bool {
-		this getEntry(key, null)
+		hashkey := This hash(key) % this capacity
+		entry := this _buckets[hashkey]
+		while (entry != null && !this _keyEquals(key, entry _key))
+			entry = entry _next
+		entry != null
+	}
+	each: func (action: Func (V*)) {
+		for (index in 0 .. this capacity) {
+			entry := this _buckets[index]
+			while (entry != null) {
+				action(entry _value&)
+				entry = entry _next
+			}
+		}
 	}
-	remove: func (key: K) -> Bool {
-		hash : SizeT = this hashKey(key) % this capacity
-		prev: HashEntry* = null
-		entry: HashEntry* = (this buckets data as HashEntry*)[hash]&
-
-		result := false
-		if (entry@ key != null) {
-			while (true) {
-				if (This _keyEquals(entry@ key as K, key)) {
-					memfree(entry@ key)
-					memfree(entry@ value)
-
-					if (prev)
-						prev@ next = entry@ next
-					else if (entry@ next)
-						this buckets[hash] = entry@ next@
-					else
-						this buckets[hash] = HashEntry new()
-
-					for (i in 0 .. this keys count) {
-						cKey := this keys[i]
-						if (This _keyEquals(key, cKey)) {
-							this keys removeAt(i)
-							break
-						}
-					}
-					this _count -= 1
-					result = true
-					break
-				}
-				if (entry@ next) {
-					prev = entry
-					entry = entry@ next
-				} else
-					break
+	each: func ~withKeys (action: Func (K*, V*)) {
+		for (index in 0 .. this capacity) {
+			entry := this _buckets[index]
+			while (entry != null) {
+				action(entry _key&, entry _value&)
+				entry = entry _next
 			}
 		}
-		result
 	}
-	resize: func (_capacity: SizeT) {
+	resize: func (newCapacity: Int) {
```

Comment:
Note that it's the number of buckets, not the number of elements the map can store (which is "unlimited").